### PR TITLE
Have radio groups work with boolean values

### LINKED ISSUE #1855: [BUG] Radio Group trouble with booleans (Draft solution included!)
- [X] I have written a [good issue](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Issue) 

## Describe the bug
When two [`Kirby-radio`](https://cookbook.kirby.design/#/home/showcase/radio) elements are used in the scenario where: 
- They are in a `kirby-radio-group` 
- One of the radios has `[value]="true"` and the other has `[value]="false` 
- The `kirby-radio-group` element has its initial value set to `false` via `[value]="false"`

The corresponding radio button will not be checked as expected. If instead the `kirby-radio-group` element has its initial value set to `true` via `[value]="true"` the corresponding radio will be checked. 

## Describe how to reproduce the bug
I've created a snippet that can be used to recreate this behaviour in the cookbook: 
```Typescript 
import { Component } from '@angular/core';

const config = {
  selector: 'cookbook-radio-example-binding',
  template: `
  <kirby-radio-group [value]="myBoolVar" (valueChange)="onChange($event)">
    <kirby-radio [value]="true" [text]="'yes'"></kirby-radio>
    <kirby-radio [value]="false" [text]="'no'"></kirby-radio>
  </kirby-radio-group>
`,
};

@Component({
  selector: config.selector,
  template: config.template,
})
export class RadioExampleBindingComponent {
  myBoolVar = false;
  onChange(item: { title: string; value: number }) {
    console.dir(item);
  }
}
```

## Which Kirby version was used?
4.0.10

## What was the expected behavior?
The radio button with value `false` should be checked initially. 

## Please complete the following information:
- OS: MacOS
- Browser tested in Chrome, Safari, Firefox

## Are there any additional context?
I've created a draft solution to this that needs some testing.

<hr />

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Verification
To make sure the bug is not intended behaviour; it should be verified by a member of team Kirby before moving on to implementation. 

- [ ] Make sure the `NOT verified` label has been removed by a member of [team Kirby](https://github.com/kirbydesign/designsystem/blob/master/.github/SUPPORT.md#team-kirby); do not proceed until this is done

### Implementation 
The contributor who wants to implement this issue should: 

- [ ] Make sure you have read: "[Before you get coding](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#before-you-get-coding)".
- [ ] Signal to others you are working on the issue by assigning yourself.
- [ ] Create a branch from the [master branch](https://github.com/kirbydesign/designsystem/tree/master) following our [branch naming convention](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Branch). 
- [ ] Create a test that reproduces the bug following guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)". 
- [ ] Write code to fix the bug

### Review
Once the issue has been implemented and is ready for review:

- [ ] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review). 
- [ ] Create a pull-request.

### PR DESCRIPTION
**Pull request by MadsBuchmann**
## Which issue does this PR close?

This PR closes #1855

## What is the new behavior?

The radio group component can now work with boolean values. 
The problem stemmed from the assessments in the radio group component treating `undefined` and `false` the same; while `false` is a valid value for a radio group. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Are there any additional context?

This probably needs some more testing before we merge it. 

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Reminders
- [ ] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [ ] Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).

### Review  
- [ ] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [ ] Request that the changes are code-reviewed 
- [ ] Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)

When the pull request has been approved it will be automatically merged to master via [automerge](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request).

### CODE REVIEW

File: libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.ts [hunk 1]
```diff
@@ -236,10 +236,15 @@ export class RadioGroupComponent implements AfterContentInit, ControlValueAccess
...
+    if (this.value || this.value === false) {
       this._selectedIndex = this.getIndexOfSelectedValue(); // Ensure selectedIndex reflects value within items
     }
-    this._value = this.getValueFromSelectedIndex() || null; // Ensure value exists within items
+
+    const valueFromSelectedIndex = this.getValueFromSelectedIndex();
+    this._value =
+      !valueFromSelectedIndex && valueFromSelectedIndex !== false
+        ? valueFromSelectedIndex || null
+        : valueFromSelectedIndex; // Ensure value exists within items
```
Comment (jkaltoft) [inline libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.ts:247]: I find this logically weird to look at :)

In a simplified form it reads:

_Value is equal to either A or null or A_

The expression `!valueFromSelectedIndex && valueFromSelectedIndex !== false` is also a bit odd - probably due to the fact that we're trying to test if `valueFromSelectedIndex` is falsy but not false.

The same goes for the remaining two changes in this commit.

I suggest we huddle up and put "hovederne i blød" - maybe dust off the old boolean algebra and Demorgan's laws 😁

Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.ts:247]: Lol had to spend ~30 seconds too long to mentally parse my own code again. I will take a look at it 😅

Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.ts:247]: I must have had a stroke or something when I wrote it. Here's something that hopefully communicates intent much clearer! 

``` Typescript   
this._value =
      valueFromSelectedIndex || typeof valueFromSelectedIndex === 'boolean'
        ? valueFromSelectedIndex
        : null; // Ensure value exists within items
```

Comment (RasmusKjeldgaard) [reply libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.ts:247]: I like it, maybe with a comment for later detailing why these checks are necessary, so we dont have to guess the reasoning next time, like you had to this time around.

Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.ts]: I've refactored the bug fix. I think it's a lot cleaner now and intuition shines through more clearly.

**Review by jkaltoft — COMMENTED** [review]
Much easier to read :)

I agree with @RasmusKjeldgaard's comments about adding a comment and tests

File: libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts [hunk 2]
```diff
@@ -457,7 +488,8 @@ fdescribe('RadioGroupComponent', () => {
                     const newItems =
                       dataScenario.type === DataScenarioTypes.TEXT
                         ? ['New Guy'].concat(textItems)
-                        : [{ text: 'New Guy', value: 10 }].concat(dataItems);
+                        : /* @ts-ignore */
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts:491]: handle this ts-ignore

File: libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts [hunk 3]
```diff
@@ -193,13 +207,18 @@ describe('RadioGroupComponent', () => {
...
-
-                  expect(radioChecked(0)).toBeFalse();
-                  expect(radioChecked(1)).toBeFalse();
-                  expect(radioChecked(2)).toBeTrue();
+                  await TestHelper.whenTrue(() => radioChecked(0));
+
+                  expect(dataScenario.items.length).toEqual(ionRadioElements.length);
+                  ionRadioElements.forEach((_, index) => {
+                    if (index === 0) {
+                      expect(radioChecked(0)).toBeTrue();
```
Comment (jkaltoft) [inline libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts:217]:
```suggestion
                      expect(radioChecked(index)).toBeTrue();
```

File: libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts [hunk 4]
```diff
@@ -193,13 +207,18 @@ describe('RadioGroupComponent', () => {
...
-                  await TestHelper.whenTrue(() => radioChecked(2));
-
-                  expect(radioChecked(0)).toBeFalse();
-                  expect(radioChecked(1)).toBeFalse();
-                  expect(radioChecked(2)).toBeTrue();
+                  await TestHelper.whenTrue(() => radioChecked(0));
+
+                  expect(dataScenario.items.length).toEqual(ionRadioElements.length);
+                  ionRadioElements.forEach((_, index) => {
+                    if (index === 0) {
```
Comment (jkaltoft) [inline libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts:216]: Consider using a constant for the comparison similar to what you do in e.g. `'should have selected radio corresponding to the selected data item'` and `'should update the selected radio when the bound field is updated'`

File: libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts [hunk 5]
```diff
@@ -596,22 +646,28 @@ describe('RadioGroupComponent', () => {
                     });
 
                     it('should have correct new selected item', async () => {
+                      const newSelectedIndex = 0;
                       spectator.setInput('selectedIndex', newSelectedIndex);
                       // Wait for radio checked attribute to be updated;
-                      await TestHelper.whenTrue(() => radioChecked(2));
+                      await TestHelper.whenTrue(() => radioChecked(0));
```
Comment (jkaltoft) [inline libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts:652]:
```suggestion
                      await TestHelper.whenTrue(() => radioChecked(newSelectedIndex));
```

File: libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts (hunk 4 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.spec.ts:216]: Changed 214 into:
```typescript 
expect(dataScenario.items).not.toHaveLength(0);```

**Review by jkaltoft — APPROVED** [review]

**Pull request merged**